scrapy/core/spidermw.py

```diff
@@ -50,21 +55,36 @@ def process_spider_input(response):
 
         def process_spider_exception(_failure):
             exception = _failure.value
+            # ignore AssertionError from middleware's manager
+            if isinstance(exception, AssertionError) and len(exception.args) > 0 \
```

Comment:
I think a better way would be to raise a special scrapy-specific exception type, not AssertionError, and ignore it instead. This code could silently ignore AssertionError raised in user code.